applications/nrf_desktop/src/modules/led_stream.c

```diff
@@ -0,0 +1,250 @@
+
+	struct led_event *led_event = new_led_event();
+
+	led_event->stream = true;
+	led_event->led_id = led->id;
+	led_event->led_effect = effect;
+
+	__ASSERT_NO_MSG(led_event->led_effect->steps);
+
+	EVENT_SUBMIT(led_event);
+}
+
+static size_t count_free_places(struct led *led)
+{
+	size_t len = (CONFIG_DESKTOP_LED_STREAM_QUEUE_SIZE + led->rx_idx - led->tx_idx)
+		     % CONFIG_DESKTOP_LED_STREAM_QUEUE_SIZE;
+	return CONFIG_DESKTOP_LED_STREAM_QUEUE_SIZE - len;
+}
+
+static bool store_data(const struct event_dyndata *data, struct led *led)
+{
+	size_t free_places = count_free_places(led);
+
+	if (free_places > 0) {
+		LOG_INF("Insert data on position %" PRIu8
+			", free places %zu", led->rx_idx, free_places);
+
+		if (!queue_data(data, led)) {
+			return false;
+		}
+	} else {
+		LOG_WRN("Queue is full - dropping incoming step");
+	}
+	return true;
+}
+
+static void send_data_from_queue(struct led *led)
+{
+	size_t free_places = count_free_places(led);
+
+	if (free_places < CONFIG_DESKTOP_LED_STREAM_QUEUE_SIZE) {
+		LOG_INF("Sending data");
+
+		struct led_effect *lse = &led->led_stream_effect[led->used_effect];
+		led->used_effect = !led->used_effect;
+
+		lse->steps = &led->steps_queue[led->tx_idx];
+		lse->step_count = 1;
+
+		send_effect(lse, led);
+
+		led->tx_idx = next_index(led->tx_idx);
+	} else {
+		LOG_INF("No steps ready in queue");
+		led->first_sent = false;
+	}
+}
+
+static void handle_incoming_step(const struct config_event *event)
+{
+	if (!initialized) {
+		LOG_WRN("Not initialized");
+		return;
+	}
+
+	size_t led_id = event->dyndata.data[LED_ID_POS];
+	struct led *led = &leds[led_id];
+	led->id = led_id;
+
+
+	if (!store_data(&event->dyndata, led)) {
+		return;
+	}
+
+	if (!led->first_sent) {
+		LOG_INF("Sending first led effect");
+
+		send_data_from_queue(led);
+		led->first_sent = true;
+	}
+}
+
+static void handle_config_event(const struct config_event *event)
+{
+	if (!event->store_needed) {
+		/* Accept only events coming from transport. */
+		return;
+	}
+
+	if (GROUP_FIELD_GET(event->id) != EVENT_GROUP_LED_STREAM) {
+		/* Only LED STREAM events. */
+		return;
+	}
+
+	switch (TYPE_FIELD_GET(event->id)) {
+	case STREAM_DATA:
+		handle_incoming_step(event);
+		break;
+
+	default:
+		LOG_WRN("Unknown LED STREAM config event");
+		break;
+	}
+}
+
+static void handle_config_fetch_request_event(
+	const struct config_fetch_request_event *event)
+{
+	LOG_INF("Handle config fetch request event");
+
+	if (GROUP_FIELD_GET(event->id) != EVENT_GROUP_LED_STREAM) {
+		/* Only LED STREAM events. */
+		return;
+	}
+
+	struct led *led = &leds[STREAM_LED_ID];
+
+	u8_t free_places = count_free_places(led);
+
+	LOG_INF("Free places left: %" PRIu8, free_places);
+
+	struct config_fetch_event *fetch_event =
+		new_config_fetch_event(sizeof(free_places));
+
+	fetch_event->id = event->id;
+	fetch_event->recipient = event->recipient;
+	fetch_event->channel_id = event->channel_id;
+
+	fetch_event->dyndata.data[0] = free_places;
```

Comment:
maybe we could also pass information about the `initialized` so tool knows that there is no point in writing the stream